File: android/java/v2/AmpliApp/app/src/test/java/com/example/ampliapp/AmpliTest.java
package com.example.ampliapp;

import static org.junit.Assert.assertEquals;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;

import android.content.Context;

import com.amplitude.ampli.Ampli;
import com.amplitude.ampli.EventWithAllProperties;
import com.amplitude.ampli.Group;
import com.amplitude.ampli.Identify;
import com.amplitude.ampli.LoadClientOptions;
import com.amplitude.ampli.LoadOptions;
import com.amplitude.android.Amplitude;
import com.amplitude.android.events.BaseEvent;
import com.amplitude.android.events.EventOptions;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.junit.MockitoJUnitRunner;

import java.util.Map;

@RunWith(MockitoJUnitRunner.class)
public class AmpliTest {
    Ampli ampli;

    private final String userId = "test-ampli-user-id";
    private final String deviceId = "test-ampli-device-id";

    Amplitude client = mock(Amplitude.class);
    Context appContext = mock(Context.class);
    ArgumentCaptor<BaseEvent> eventCaptor = ArgumentCaptor.forClass(BaseEvent.class);
    ArgumentCaptor<EventOptions> optionsCaptor = ArgumentCaptor.forClass(EventOptions.class);
    ArgumentCaptor<String[]> stringArrayCaptor = ArgumentCaptor.forClass(String[].class);
    ArgumentCaptor<Map<String, Object>> mapCaptor = ArgumentCaptor.forClass(Map.class);

    @Before
    public void setUp() {
        this.ampli = new Ampli();
    }

    @Test
    public void load() {
        this.ampli.load(appContext, new LoadOptions().setClient(new LoadClientOptions().setInstance(client)));

        verify(client, times(0)).build();
    }

    @Test
    public void identify() {
        this.ampli.load(appContext, new LoadOptions().setClient(new LoadClientOptions().setInstance(client)));
        EventOptions eventOptions = new EventOptions();
        eventOptions.setDeviceId(deviceId);
        eventOptions.setUserId(userId);
        this.ampli.identify(
                userId,
                Identify.builder().requiredNumber(42.0).optionalArray(new String[]{"A", "ray"}).build(),
                eventOptions
        );

        verify(client, times(1)).identify(mapCaptor.capture(), optionsCaptor.capture());
        assertEquals(
                "{\"requiredNumber\":42,\"optionalArray\":[\"A\",\"ray\"]}",
                new JSONObject(mapCaptor.getValue()).toString()
        );
        assertEquals(
                userId,
                optionsCaptor.getValue().getUserId()
        );
        assertEquals(
                deviceId,
                optionsCaptor.getValue().getDeviceId()
        );
    }

    @Test
    public void setGroupSingleValue() {
        this.ampli.load(appContext, new LoadOptions().setClient(new LoadClientOptions().setInstance(client)));

        EventOptions eventOptions = new EventOptions();
        eventOptions.setDeviceId(deviceId);
        eventOptions.setUserId(userId);
        this.ampli.setGroup(
                "group-1",
                "value-1",
                eventOptions
        );

        verify(client, times(1)).setGroup(eq("group-1"), eq("value-1"), optionsCaptor.capture());

        assertEquals(
                userId,
                optionsCaptor.getValue().getUserId()
        );
        assertEquals(
                deviceId,
                optionsCaptor.getValue().getDeviceId()
        );
    }

 @Test
    public void setGroupMultipleValue() throws JSONException {
        this.ampli.load(appContext, new LoadOptions().setClient(new LoadClientOptions().setInstance(client)));

        EventOptions eventOptions = new EventOptions();
        eventOptions.setDeviceId(deviceId);
        eventOptions.setUserId(userId);
        this.ampli.load(appContext, new LoadOptions().setClient(new LoadClientOptions().setInstance(client)));

        this.ampli.setGroup(
                "group-1",
                new String[]{"value-1", "value-2", "value-3"},
                eventOptions
        );


        verify(client, times(1)).setGroup(eq("group-1"), stringArrayCaptor.capture(), optionsCaptor.capture());

         assertEquals(
                "[\"value-1\",\"value-2\",\"value-3\"]",
                 new JSONArray(stringArrayCaptor.getValue()).toString()
        );
        assertEquals(
             userId,
             optionsCaptor.getValue().getUserId()
        );
        assertEquals(
             deviceId,
             optionsCaptor.getValue().getDeviceId()
        );
 }

    @Test
    public void groupIdentify() {
        this.ampli.load(appContext, new LoadOptions().setClient(new LoadClientOptions().setInstance(client)));

        EventOptions eventOptions = new EventOptions();
        eventOptions.setDeviceId(deviceId);
        eventOptions.setUserId(userId);
        this.ampli.groupIdentify(
                "group-type-1",
                "group-name-1",
                Group.builder().requiredBoolean(false).optionalString("test-string").build(),
                eventOptions
        );

       verify(client, times(1)).groupIdentify(eq("group-type-1"), eq("group-name-1"), mapCaptor.capture(), optionsCaptor.capture());

        assertEquals(
                "{\"requiredBoolean\":false,\"optionalString\":\"test-string\"}",
                new JSONObject(mapCaptor.getValue()).toString()
        );
        assertEquals(
                userId,
                optionsCaptor.getValue().getUserId()
        );
        assertEquals(
                deviceId,
                optionsCaptor.getValue().getDeviceId()
        );
    }

    @Test
    public void trackEventNoProperties() {
        Amplitude client = mock(Amplitude.class);
        Context appContext = mock(Context.class);
        this.ampli.load(appContext, new LoadOptions().setClient(new LoadClientOptions().setInstance(client)));

        EventOptions eventOptions = new EventOptions();
        eventOptions.setDeviceId(deviceId);
        eventOptions.setUserId(userId);
        this.ampli.eventNoProperties(eventOptions);

        verify(client, times(1)).track(
                eventCaptor.capture(),
                optionsCaptor.capture()
        );
        assertEquals(
                userId,
                optionsCaptor.getValue().getUserId()
        );
        assertEquals(
                deviceId,
                optionsCaptor.getValue().getDeviceId()
        );
        assertEquals(
                "Event No Properties",
                 eventCaptor.getValue().getEventType()
        );
        assertEquals("{}", new JSONObject(eventCaptor.getValue().getEventProperties()).toString());
    }

    @Test
    public void trackEventWithAllProperties() {
        Amplitude client = mock(Amplitude.class);
        Context appContext = mock(Context.class);
        this.ampli.load(appContext, new LoadOptions().setClient(new LoadClientOptions().setInstance(client)));

        EventOptions eventOptions = new EventOptions();
        eventOptions.setDeviceId(deviceId);
        eventOptions.setUserId(userId);
        this.ampli.eventWithAllProperties(
                EventWithAllProperties.builder()
                        .requiredArray(new String[]{"Required", "array"})
                        .requiredBoolean(true)
                        .requiredEnum(EventWithAllProperties.RequiredEnum.ENUM2)
                        .requiredInteger(41)
                        .requiredNumber(42.0)
                        .requiredString("Required string")
                        .build(),
                eventOptions
        );

        verify(client, times(1)).track(
                eventCaptor.capture(),
                optionsCaptor.capture()
        );

        assertEquals(
                userId,
                optionsCaptor.getValue().getUserId()
        );
        assertEquals(
                deviceId,
                optionsCaptor.getValue().getDeviceId()
        );
        assertEquals(
                "Event With All Properties",
                eventCaptor.getValue().getEventType()
        );
        assertEquals(
                "{\"requiredEnum\":\"Enum2\",\"requiredArray\":[\"Required\",\"array\"],\"requiredNumber\":42,\"requiredConst\":\"some-const-value\",\"requiredBoolean\":true,\"requiredInteger\":41,\"requiredString\":\"Required string\"}",
                new JSONObject(eventCaptor.getValue().getEventProperties()).toString()
        );
    }

    @Test
    public void flush() {
        Amplitude client = mock(Amplitude.class);
        Context appContext = mock(Context.class);
        this.ampli.load(appContext, new LoadOptions().setClient(new LoadClientOptions().setInstance(client)));

        this.ampli.flush();

        verify(client, times(1)).flush();
    }
}

File: jre/kotlin/AmpliApp/src/test/kotlin/AmpliTest.kt
package com.example.ampliapp

import com.amplitude.ampli.*
import com.amplitude.Amplitude
import com.amplitude.Event
import com.amplitude.MiddlewareExtra
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import kotlin.test.assertEquals
import kotlin.test.Test
import kotlin.test.BeforeTest
import org.mockito.ArgumentCaptor
import org.mockito.Captor
import org.mockito.Mockito.*
import kotlin.test.assertNull

class AmpliTest {
    lateinit var ampli: Ampli

    private val userId = "test-ampli-user-id"
    private val deviceId = "test-ampli-device-id"

    @Captor
    lateinit var eventCaptor: ArgumentCaptor<Event>
    @Captor
    lateinit var extraCaptor: ArgumentCaptor<MiddlewareExtra>

    @BeforeTest
    fun setUp() {
        this.ampli = Ampli()
        this.eventCaptor = ArgumentCaptor.forClass(Event::class.java)
        this.extraCaptor = ArgumentCaptor.forClass(MiddlewareExtra::class.java)
    }

    @Test
    fun load() {
        val client = mock(Amplitude::class.java)

        this.ampli.load(LoadOptions(client = LoadClientOptions(instance = client, apiKey = "sample-api-key")))

        verify(client, times(0)).init(any())
    }

    @Test
    fun identify() {
        val client = mock(Amplitude::class.java)
        this.ampli.load(LoadOptions(client = LoadClientOptions(instance = client)))

        val extra = MiddlewareExtra(mapOf("abc" to 123, "xyz" to "987"))

        this.ampli.identify(
            userId,
            Identify(
                requiredNumber = 42.0,
                optionalArray = arrayOf("A", "ray")
            ).options(deviceId = deviceId),
            EventOptions(userId = "some-user"),
            extra
        )

        verify(client, times(1)).logEvent(eventCaptor.capture(), extraCaptor.capture())

        val event = eventCaptor.value
        assertEquals("\$identify", event.eventType)
        assertEquals(userId, event.userId)
        assertEquals(deviceId, event.deviceId)
        assertEquals(
            """{
  "requiredNumber": 42,
  "optionalArray": [
    "A",
    "ray"
  ]
}""", event.userProperties.toString(2)
        )
        assertEquals("{xyz=987, abc=123}", extraCaptor.value.toString())
    }

    @Test
    fun setGroupSingleValue() {
        val client = mock(Amplitude::class.java)
        this.ampli.load(LoadOptions(client = LoadClientOptions(instance = client)))

        val extra = MiddlewareExtra(mapOf("abc" to 123, "xyz" to "987"))

        val groupProperties = JSONObject()
        try {
            groupProperties.put("group-1", "value-1")
        } catch (e: JSONException) {
            System.err.println("Error converting properties to JSONObject: ${e.message}")
        }

        val amplitudeEvent = Event("\$identify", userId)
        amplitudeEvent.groupProperties = groupProperties
        this.ampli.client.logEvent(amplitudeEvent, extra)

        verify(client, times(1)).logEvent(eventCaptor.capture(), extraCaptor.capture())

        val event = eventCaptor.value
        assertEquals("\$identify", event.eventType)
        assertEquals(userId, event.userId)
        assertEquals(
            """{"group-1": "value-1"}""", event.groupProperties.toString(2)
        )
        assertEquals("{xyz=987, abc=123}", extraCaptor.value.toString())
    }

    @Test
    fun setGroupMultipleValues() {
        val client = mock(Amplitude::class.java)
        this.ampli.load(LoadOptions(client = LoadClientOptions(instance = client)))

        val extra = MiddlewareExtra(mapOf("abc" to 123, "xyz" to "987"))

        val groupNames = try {
            JSONArray(arrayOf("value-1", "value-2", "value-3"))
        } catch (e: JSONException) {
            System.err.printf("Error converting value to JSONArray: %s%n", e.message)
            JSONArray()
        }

        val groupProperties = JSONObject()
        try {
            groupProperties.put("group-1", groupNames)
        } catch (e: JSONException) {
            System.err.println("Error converting properties to JSONObject: ${e.message}")
        }

        val amplitudeEvent = Event("\$identify", userId)
        amplitudeEvent.groupProperties = groupProperties

        this.ampli.client.logEvent(amplitudeEvent, extra)

        verify(client, times(1)).logEvent(eventCaptor.capture(), extraCaptor.capture())

        val event = eventCaptor.value
        assertEquals("\$identify", event.eventType)
        assertEquals(userId, event.userId)
        assertEquals(
            """{"group-1": [
  "value-1",
  "value-2",
  "value-3"
]}""", event.groupProperties.toString(2)
        )
        assertEquals("{xyz=987, abc=123}", extraCaptor.value.toString())
    }

    @Test
    fun trackEventNoProperties() {
        val client = mock(Amplitude::class.java)
        this.ampli.load(LoadOptions(client = LoadClientOptions(instance = client)))

        val extra = MiddlewareExtra(mapOf("abc" to 123, "xyz" to "987"))

        this.ampli.eventNoProperties(userId, EventOptions(deviceId = deviceId, userId = "some-user"), extra)

        verify(client, times(1)).logEvent(eventCaptor.capture(), extraCaptor.capture())

        val event = eventCaptor.value
        assertEquals("Event No Properties", event.eventType)
        assertEquals(userId, event.userId)
        assertEquals(deviceId, event.deviceId)
        assertNull(event.eventProperties)
        assertEquals("{xyz=987, abc=123}", extraCaptor.value.toString())
    }

    @Test
    fun trackEventWithAllProperties() {
        val client = mock(Amplitude::class.java)
        this.ampli.load(LoadOptions(client = LoadClientOptions(instance = client)))

        val extra = MiddlewareExtra(mapOf("abc" to 123, "xyz" to "987"))

        this.ampli.track(null,
            EventWithAllProperties(
                requiredString = "Required string",
                requiredArray = arrayOf("Required", "array"),
                requiredBoolean = true,
                requiredEnum = EventWithAllProperties.RequiredEnum.ENUM_2,
                requiredInteger = 41,
                requiredNumber = 42.0
            ).options(EventOptions(deviceId = deviceId)),
            EventOptions(userId = userId),
            extra
        )

        verify(client, times(1)).logEvent(eventCaptor.capture(), extraCaptor.capture())

        val event = eventCaptor.value
        assertEquals("Event With All Properties", event.eventType)
        assertEquals(userId, event.userId)
        assertEquals(deviceId, event.deviceId)
        assertEquals(
            """{
  "requiredEnum": "Enum2",
  "requiredArray": [
    "Required",
    "array"
  ],
  "requiredNumber": 42,
  "requiredConst": "some-const-value",
  "requiredBoolean": true,
  "requiredInteger": 41,
  "requiredString": "Required string"
}""", event.eventProperties.toString(2)
        )
        assertEquals("{xyz=987, abc=123}", extraCaptor.value.toString())
    }

    @Test
    fun flush() {
        val client = mock(Amplitude::class.java)
        this.ampli.load(LoadOptions(client = LoadClientOptions(instance = client)))

        this.ampli.flush()

        verify(client, times(1)).flushEvents()
    }

    @Test
    fun eventOptionsSessionId() {
        val client = mock(Amplitude::class.java)
        this.ampli.load(LoadOptions(client = LoadClientOptions(instance = client)))
        this.ampli.eventNoProperties(userId, EventOptions(deviceId = deviceId, userId = "some-user", sessionId = 999L))
        verify(client, times(1)).logEvent(eventCaptor.capture(), extraCaptor.capture())
        val event = eventCaptor.value
        assertEquals(999L, event.sessionId)
    }
}